flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java

```diff
@@ -980,110 +845,41 @@ protected ActorGateway getJobManagerGateway(CommandLineOptions options) throws E
-			}
-
-			// in case the main detached mode wasn't set, we don't wanna overwrite the one loaded
-			// from yarn options.
-			if (detachedMode) {
-				flinkYarnClient.setDetachedMode(true);
-			}
-
-			// the number of slots available from YARN:
-			int yarnTmSlots = flinkYarnClient.getTaskManagerSlots();
-			if (yarnTmSlots == -1) {
-				yarnTmSlots = 1;
-			}
-			maxSlots = yarnTmSlots * flinkYarnClient.getTaskManagerCount();
-			if (userParallelism != -1) {
-				int slotsPerTM = userParallelism / flinkYarnClient.getTaskManagerCount();
-				logAndSysout("The YARN cluster has " + maxSlots + " slots available, " +
-						"but the user requested a parallelism of " + userParallelism + " on YARN. " +
-						"Each of the " + flinkYarnClient.getTaskManagerCount() + " TaskManagers " +
-						"will get "+slotsPerTM+" slots.");
-				flinkYarnClient.setTaskManagerSlots(slotsPerTM);
-			}
-
-			try {
-				yarnCluster = flinkYarnClient.deploy();
-				yarnCluster.connectToCluster();
-			}
-			catch (Exception e) {
-				throw new RuntimeException("Error deploying the YARN cluster", e);
-			}
+		// Get the custom command-line (e.g. Standalone/Yarn/Mesos)
+		CustomCommandLine<?> activeCommandLine =
+			CliFrontendParser.getActiveCustomCommandLine(options.getJobManagerAddress());
 
-			jobManagerAddress = yarnCluster.getJobManagerAddress();
-			writeJobManagerAddressToConfig(jobManagerAddress);
-			
-			// overwrite the yarn client config (because the client parses the dynamic properties)
-			this.config.addAll(flinkYarnClient.getFlinkConfiguration());
-
-			logAndSysout("YARN cluster started");
-			logAndSysout("JobManager web interface address " + yarnCluster.getWebInterfaceURL());
-			logAndSysout("Waiting until all TaskManagers have connected");
-
-			while(true) {
-				GetClusterStatusResponse status = yarnCluster.getClusterStatus();
-				if (status != null) {
-					if (status.numRegisteredTaskManagers() < flinkYarnClient.getTaskManagerCount()) {
-						logAndSysout("TaskManager status (" + status.numRegisteredTaskManagers() + "/"
-							+ flinkYarnClient.getTaskManagerCount() + ")");
-					} else {
-						logAndSysout("All TaskManagers are connected");
-						break;
-					}
-				} else {
-					logAndSysout("No status updates from the YARN cluster received so far. Waiting ...");
-				}
+		ClusterClient client = activeCommandLine.retrieveCluster(options.getCommandLine(), config);
 
-				try {
-					Thread.sleep(500);
-				}
-				catch (InterruptedException e) {
-					LOG.error("Interrupted while waiting for TaskManagers");
-					System.err.println("Thread is interrupted");
-					Thread.currentThread().interrupt();
-				}
-			}
-		}
-		else {
-			if(options.getJobManagerAddress() != null) {
-				jobManagerAddress = ClientUtils.parseHostPortAddress(options.getJobManagerAddress());
-				writeJobManagerAddressToConfig(jobManagerAddress);
+		if (client != null) {
+			logAndSysout("Cluster retrieved");
+		} else {
```

Comment:
Yes, I have thought about letting the command-line decide whether a cluster can be created or retrieved (it would return an Enum based on the options supplied). I decided I would address this in a follow-up.